### ISSUE TITLE
[meta] stable/unstable thoughts

### ISSUE THREAD

**Issue by darinkishore** [lilyinstarlight/nixos-cosmic#221]
Could we make both stable and unstable supported by just using a rust overlay? (https://github.com/oxalica/rust-overlay)

It's working on my end (I'm running unstable, and 1.79 is not in unstable yet), but I'm still having some local package building problems, unsure why.

**Comment by darinkishore** [comment]
What I did:

`flake.nix`
```nix
{
  # COSMIC flake
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";

    flake-compat = {
      url = "github:nix-community/flake-compat";
      flake = false;
    };

    rust-overlay = {
      url = "github:oxalica/rust-overlay";
      inputs.nixpkgs.follows = "nixpkgs";
    };
  };

  outputs = { self, nixpkgs, rust-overlay, ... }: let
    forAllSystems = nixpkgs.lib.genAttrs [ "x86_64-linux" "aarch64-linux" "x86_64-darwin" "aarch64-darwin" ];
    
    pkgsForSystem = system: import nixpkgs {
      inherit system;
      overlays = [
        rust-overlay.overlays.default
        self.overlays.default
      ];
    };

    makeCustomRustPlatform = pkgs: let
      latestStable = rust-overlay.packages.x86_64-linux.default;
    in pkgs.makeRustPlatform {
      cargo = latestStable;
      rustc = latestStable;
    };

  in {
    lib = {
      packagesFor = system: let
        pkgs = pkgsForSystem system;
        customRustPlatform = makeCustomRustPlatform pkgs;
      in import ./pkgs { 
        inherit pkgs customRustPlatform; 
      };
    };

    packages = forAllSystems (system: 
      self.lib.packagesFor system
    );

    overlays = {
      default = final: prev: let
        customRustPlatform = makeCustomRustPlatform final;
      in import ./pkgs {
        pkgs = final;
        inherit customRustPlatform;
      };
    };

    nixosModules = {
      default = import ./nixos { cosmicOverlay = self.overlays.default; };
    };

    legacyPackages = forAllSystems (system: let
      pkgs = pkgsForSystem system;
    in {
      update = pkgs.writeShellApplication {
        name = "cosmic-update";

        runtimeInputs = [
          pkgs.coreutils
          pkgs.nix-update
        ];

        text = ''
          for pkg in pkgs/*; do
            if ! [ -f "$pkg/package.nix" ]; then
              continue
            fi

            attr="$(basename "$pkg")"

            if [ "$attr" = libcosmicAppHook ]; then
              continue
            fi

            nix-update --commit --version branch=HEAD "$attr"
          done
        '';
      };

      vm = nixpkgs.lib.makeOverridable ({ nixpkgs }: let
        nixosConfig = nixpkgs.lib.nixosSystem {
          modules = [
            ({ lib, pkgs, modulesPath, ... }: {
              imports = [
                self.nixosModules.default

                "${builtins.toString modulesPath}/virtualisation/qemu-vm.nix"
              ];

              services.desktopManager.cosmic.enable = true;
              services.displayManager.cosmic-greeter.enable = true;

              services.flatpak.enable = true;

              environment.systemPackages = [ pkgs.drm_info pkgs.firefox pkgs.cosmic-emoji-picker pkgs.cosmic-tasks ];

              boot.kernelParams = [ "quiet" "udev.log_level=3"  ];
              boot.initrd.kernelModules = [ "bochs" ];

              boot.initrd.verbose = false;

              boot.initrd.systemd.enable = true;

              boot.loader.systemd-boot.enable = true;
              boot.loader.timeout = 0;

              boot.plymouth.enable = true;
              boot.plymouth.theme = "nixos-bgrt";
              boot.plymouth.themePackages = [ pkgs.nixos-bgrt-plymouth ];

              services.openssh = {
                enable = true;
                settings.PermitRootLogin = "yes";
              };

              documentation.nixos.enable = false;

              users.mutableUsers = false;
              users.users.root.password = "meow";
              users.users.user = {
                isNormalUser = true;
                password = "meow";
              };

              virtualisation.useBootLoader = true;
              virtualisation.useEFIBoot = true;
              virtualisation.mountHostNixStore = true;

              virtualisation.memorySize = 4096;
              # TODO: below options can be removed once NixOS/nixpkgs#279009 is merged
              virtualisation.qemu.options = [ "-vga none" "-device virtio-gpu-gl-pci" "-display default,gl=on" ];

              virtualisation.forwardPorts = [
                { from = "host"; host.port = 2222; guest.port = 22; }
              ];

              nixpkgs.hostPlatform = system;

              system.stateVersion = lib.trivial.release;
            })
          ];
        };
      in nixosConfig.config.system.build.vm // { closure = nixosConfig.config.system.build.toplevel; inherit (nixosConfig) config pkgs; }) { inherit nixpkgs; };
    });
  };
}
```

`pkgs/default.nix`

```nix
{ pkgs
, customRustPlatform
}:

let
  dir = builtins.readDir ./.;
  pkgsPaths = builtins.filter (pkgPath: pkgPath != null) (
    builtins.map (
      name:
      if dir.${name} == "directory" then
        {
          inherit name;
          value = "${builtins.toString ./.}/${name}/package.nix";
        }
      else
        null
    ) (builtins.attrNames dir)
  );
  callPackage = pkgs.lib.callPackageWith (
    pkgs
    // finalPkgs
    // {
      buildPackages = finalPkgs;
      targetPackages = finalPkgs;
      inherit callPackage;
      rustPlatform = customRustPlatform;
    }
  );
  finalPkgs = builtins.listToAttrs (
    builtins.map (pkgPath: {
      inherit (pkgPath) name;
      value = callPackage pkgPath.value {};
    }) pkgsPaths
  );
in
finalPkgs
```

tl;dr: create custom rust platform based on rust-overlay's default rust version.

**Comment by drakon64** [comment]
I'm not sure this is entirely necessary for Unstable since newer Rust versions will eventually arrive there, but for Stable I can see the benefit at least until COSMIC itself standardises on a minimum supported Rust version.

**Comment by drakon64** [comment]
I'm currently attempting a build using your suggested changes, but these tend to take a few hours.

I do have one concern. This line:
```nix
latestStable = rust-overlay.packages.x86_64-linux.default;
```

suggests to me that the Rust overlay will only ever work for x86_64 while we currently have a build cache for x86_64 and AArch64. Is there a way to support both?

**Comment by darinkishore** [comment]
> I'm currently attempting a build using your suggested changes, but these tend to take a few hours.
> 
> I do have one concern. This line:
> 
> ```nix
> latestStable = rust-overlay.packages.x86_64-linux.default;
> ```
> 
> suggests to me that the Rust overlay will only ever work for x86_64 while we currently have a build cache for x86_64 and AArch64. Is there a way to support both?

Ah. Sorry about that! I don't think that was the case, I only have an x86 linux machine to hack on, so that was what I built for as I tried to get things working. 

That being said, give these a shot?

1. Modify `makeCustomRustPlatform` to use the current system:

```nix
makeCustomRustPlatform = system: pkgs: let
  latestStable = rust-overlay.packages.${system}.default;
in pkgs.makeRustPlatform {
  cargo = latestStable;
  rustc = latestStable;
};
```

2. Update `lib.packagesFor` function to pass the system to makeCustomRustPlatform:

```nix
lib = {
  packagesFor = system: let
    pkgs = pkgsForSystem system;
    customRustPlatform = makeCustomRustPlatform system pkgs;
  in import ./pkgs { 
    inherit pkgs customRustPlatform; 
  };
};
```

3. Update `overlays.default` to pass current system:

```nix
overlays = {
  default = final: prev: let
    system = final.stdenv.hostPlatform.system;
    customRustPlatform = makeCustomRustPlatform system final;
  in import ./pkgs {
    pkgs = final;
    inherit customRustPlatform;
  };
};

```

And maybe for VM, (forget if necessary/source flake does this)

```
vm = nixpkgs.lib.makeOverridable ({ nixpkgs }: let
  system = nixpkgs.system; # added
  nixosConfig = nixpkgs.lib.nixosSystem {
    modules = [
      ({ lib, pkgs, modulesPath, ... }: {
        # ... 
        nixpkgs.hostPlatform = system; # unchanged but this should use right sys, 
        # forgot if main flake built VM for all systems?? 
      })
    ];
  };
in nixosConfig.config.system.build.vm // { closure = nixosConfig.config.system.build.toplevel; inherit (nixosConfig) config pkgs; }) { inherit nixpkgs; };
```

Still fairly new to nix (<1.5 months), and I haven't actually tried any of the code, I just thiiinnnkkk these should be the right changes. 

As far as testing goes, it took me a while for builds too, which got frustrating, I started testing just one package at a time. `cosmic-settings` is probably the one to test builds for.

**Comment by darinkishore** [comment]
Closing b/c stable support removed.
